Jellyfin.Api/Controllers/LiveTvController.cs

```diff
@@ -1220,11 +1220,11 @@ public async Task<ActionResult> GetLiveStreamFile([FromRoute, Required] string s
                 return NotFound();
             }
 
-            await using var memoryStream = new MemoryStream();
             await new ProgressiveFileCopier(liveStreamInfo, null, _transcodingJobHelper, CancellationToken.None)
-                .WriteToAsync(memoryStream, CancellationToken.None)
+                .WriteToAsync(Response.Body, CancellationToken.None)
                 .ConfigureAwait(false);
-            return File(memoryStream, MimeTypes.GetMimeType("file." + container));
+            Response.ContentType = MimeTypes.GetMimeType("file." + container);
+            return Ok();
```

Comment:
I'm not 100% sure why this change is needed, but the instant I made this change playback started working.

Reply:
You have to be careful about writing directly to Response.Body and then returning `Ok` or similar methods that set the response headers.

Would #4213 work for this?

Reply:
I don't see how I could use the `ProgressiveFileStream` without a rewrite of how ts livestreams are handled, but I found that if I flush the `MemoryStream` I am still able to get the stream to play.

Reply:
How is it different? Isn't it supposed to copy from one continuous stream to the response?

And won't a memorystream cause high mem usage?

Reply:
Just to add - there was a issue logged a whilst back about an OOM error due to high-def video streaming, caused by the file being copied to a memorystream and then streamed.

Reply:
I sent this to crobibero already, but for posterity, here's a POC:

```
Index: Jellyfin.Api/Controllers/LiveTvController.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Jellyfin.Api/Controllers/LiveTvController.cs	(revision 5d88e610768c8b680096938a372585638ef18fe3)
+++ Jellyfin.Api/Controllers/LiveTvController.cs	(date 1604445766581)
@@ -1220,11 +1220,8 @@
                 return NotFound();
             }
 
-            await new ProgressiveFileCopier(liveStreamInfo, null, _transcodingJobHelper, CancellationToken.None)
-                .WriteToAsync(Response.Body, CancellationToken.None)
-                .ConfigureAwait(false);
-            Response.ContentType = MimeTypes.GetMimeType("file." + container);
-            return Ok();
+            var liveStream = new ProgressiveFileStream(liveStreamInfo.GetFilePath(), null, _transcodingJobHelper);
+            return new FileStreamResult(liveStream, MimeTypes.GetMimeType("file." + container));
         }
 
         private void AssertUserCanManageLiveTv()
Index: Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs	(revision 5d88e610768c8b680096938a372585638ef18fe3)
+++ Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs	(date 1604427652162)
@@ -131,6 +131,11 @@
             await taskCompletionSource.Task.ConfigureAwait(false);
         }
 
+        public string GetFilePath()
+        {
+            return TempFilePath;
+        }
+
         private Task StartStreaming(UdpClient udpClient, HdHomerunManager hdHomerunManager, IPAddress remoteAddress, TaskCompletionSource<bool> openTaskCompletionSource, CancellationToken cancellationToken)
         {
             return Task.Run(async () =>
Index: Jellyfin.Api/Helpers/ProgressiveFileStream.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Jellyfin.Api/Helpers/ProgressiveFileStream.cs	(revision 5d88e610768c8b680096938a372585638ef18fe3)
+++ Jellyfin.Api/Helpers/ProgressiveFileStream.cs	(date 1604445760990)
@@ -82,20 +82,23 @@
             int totalBytesRead = 0;
             int remainingBytesToRead = count;
 
+            int newOffset = offset;
             while (remainingBytesToRead > 0)
             {
                 cancellationToken.ThrowIfCancellationRequested();
                 int bytesRead;
                 if (_allowAsyncFileRead)
                 {
-                    bytesRead = await _fileStream.ReadAsync(buffer, offset, remainingBytesToRead, cancellationToken).ConfigureAwait(false);
+                    bytesRead = await _fileStream.ReadAsync(buffer, newOffset, remainingBytesToRead, cancellationToken).ConfigureAwait(false);
                 }
                 else
                 {
-                    bytesRead = _fileStream.Read(buffer, offset, remainingBytesToRead);
+                    bytesRead = _fileStream.Read(buffer, newOffset, remainingBytesToRead);
                 }
 
                 remainingBytesToRead -= bytesRead;
+                newOffset += bytesRead;
+
                 if (bytesRead > 0)
                 {
                     _bytesWritten += bytesRead;
@@ -108,12 +111,13 @@
                 }
                 else
                 {
-                    if (_job == null || _job.HasExited)
+                    // If the job is null it's a live stream and will require user action to close
+                    if (_job?.HasExited ?? false)
                     {
                         break;
                     }
 
-                    await Task.Delay(100, cancellationToken).ConfigureAwait(false);
+                    await Task.Delay(50, cancellationToken).ConfigureAwait(false);
                 }
             }
 
Index: Emby.Server.Implementations/LiveTv/TunerHosts/SharedHttpStream.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Emby.Server.Implementations/LiveTv/TunerHosts/SharedHttpStream.cs	(revision 5d88e610768c8b680096938a372585638ef18fe3)
+++ Emby.Server.Implementations/LiveTv/TunerHosts/SharedHttpStream.cs	(date 1604427652166)
@@ -122,6 +122,11 @@
             }
         }
 
+        public string GetFilePath()
+        {
+            return TempFilePath;
+        }
+
         private Task StartStreaming(HttpResponseMessage response, TaskCompletionSource<bool> openTaskCompletionSource, CancellationToken cancellationToken)
         {
             return Task.Run(async () =>
Index: MediaBrowser.Controller/Library/IMediaSourceManager.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- MediaBrowser.Controller/Library/IMediaSourceManager.cs	(revision 5d88e610768c8b680096938a372585638ef18fe3)
+++ MediaBrowser.Controller/Library/IMediaSourceManager.cs	(date 1604427652170)
@@ -115,5 +115,7 @@
     public interface IDirectStreamProvider
     {
         Task CopyToAsync(Stream stream, CancellationToken cancellationToken);
+
+        string GetFilePath();
     }
 }
```